File: src/main/java/br/com/geradormassa/generators/ProdutoGenerator.java
package br.com.geradormassa.generators;

import br.com.geradormassa.model.Produto;
import br.com.geradormassa.utils.StringUtils;

import java.text.DecimalFormat;
import java.util.Random;

/**
 * Gera dados de produtos de forma local e confiável, implementando a interface Gerador.
 * Agora, toda a lógica de criação de um produto, incluindo quantidade e descrição,
 * está centralizada aqui.
 */
public class ProdutoGenerator implements Gerador<Produto> { // 1. Implementa a interface

    private static final Random random = new Random();
    // Formata o preço com ponto, para consistência.
    private static final DecimalFormat df = new DecimalFormat("0.00");

    private static final String[] NOMES_PRODUTO = {
            "Smartphone", "Notebook", "Fone de Ouvido", "Smartwatch", "Tablet", "Camera Digital",
            "Teclado Mecanico", "Mouse Gamer", "Monitor 4K", "Cadeira Gamer"
    };

    private static final String[] CATEGORIAS = {
            "Eletronicos", "Informatica", "Acessorios", "Fotografia", "Perifericos", "Games"
    };

    private static final String[] ADJETIVOS = {
            "Moderno", "Potente", "Sem Fio", "Ultra-fino", "Profissional", "RGB", "Ergonomica"
    };

    // 2. O construtor agora é público (o construtor padrão, que é público).
    public ProdutoGenerator() {}

    /**
     * Gera um objeto Produto completo com dados aleatórios e consistentes.
     * @return Um objeto {@link Produto} preenchido.
     */
    @Override // 3. Implementa o método da interface
    public Produto gerar() {
        String categoria = CATEGORIAS[random.nextInt(CATEGORIAS.length)];
        String nomeBase = NOMES_PRODUTO[random.nextInt(NOMES_PRODUTO.length)];
        String adjetivo = ADJETIVOS[random.nextInt(ADJETIVOS.length)];
        String nomeProduto = nomeBase + " " + adjetivo;

        String descricao = "Um excelente " + nomeBase.toLowerCase() + " " + adjetivo.toLowerCase() +
                " da categoria " + categoria + ", ideal para suas necessidades.";

        // Gera um preço entre 50.00 e 4999.99
        double precoDouble = 50.0 + (4950.0 * random.nextDouble());
        String preco = df.format(precoDouble).replace(",", "."); // Garante ponto como separador

        // Gera uma quantidade aleatória e a descrição da categoria
        int quantidade = random.nextInt(100) + 1;
        String descricaoCategoria = "Descricao para a categoria " + categoria;

        return Produto.builder()
                .nomeProduto(StringUtils.removerAcentos(nomeProduto))
                .preco(preco)
                .descricaoProduto(StringUtils.removerAcentos(descricao))
                .quantidade(quantidade)
                .nomeCategoria(StringUtils.removerAcentos(categoria))
                .descricaoCategoria(StringUtils.removerAcentos(descricaoCategoria))
                .build();
    }
}

File: src/generators/produtoGenerator.ts
import { faker } from '@faker-js/faker';
import { Produto } from '../models/Produto';

const NOMES_PRODUTO = [
    "Smartphone", "Notebook", "Fone de Ouvido", "Smartwatch", "Tablet", "Camera Digital",
    "Teclado Mecanico", "Mouse Gamer", "Monitor 4K", "Cadeira Gamer", "Impressora 3D",
    "Projetor HD", "SSD Externo", "Power Bank", "Roteador Wi-Fi 6", "Placa de Video",
    "Processador", "Memoria RAM", "Cooler CPU", "Gabinete ATX", "Webcam Full HD",
    "Microfone Condensador", "Mesa Digitalizadora", "Drone", "Soundbar", "Smart TV 8K",
    "Console de Videogame", "Oculos de Realidade Virtual", "Leitor de E-book", "HD Externo"
];

const CATEGORIAS = [
    "Eletronicos", "Informatica", "Acessorios", "Fotografia", "Perifericos", "Games",
    "Hardware", "Redes e Conectividade", "Audio e Video", "Componentes", "Casa Inteligente",
    "Dispositivos Moveis", "Armazenamento", "Energia", "Impressao"
];

const ADJETIVOS = [
    "Moderno", "Potente", "Sem Fio", "Ultra-fino", "Profissional", "RGB", "Ergonomica",
    "Compacto", "Resistente", "Silencioso", "Veloz", "Inteligente", "Premium", "Essencial",
    "Versatil", "Sustentavel", "Customizavel", "Portatil", "Imersivo", "Conectado"
];

function removerAcentos(texto: string): string {
    return texto.normalize("NFD").replace(/[\u0300-\u036f]/g, "");
}

export function gerarProduto(): Produto {
    const categoria = faker.helpers.arrayElement(CATEGORIAS);
    const nomeBase = faker.helpers.arrayElement(NOMES_PRODUTO);
    const adjetivo = faker.helpers.arrayElement(ADJETIVOS);
    const nomeProduto = `${nomeBase} ${adjetivo}`;

    const descricao = `Um excelente ${nomeBase.toLowerCase()} ${adjetivo.toLowerCase()} da categoria ${categoria}, ideal para suas necessidades.`;
    const preco = faker.commerce.price({ min: 50, max: 4999, dec: 2 });
    const quantidade = faker.number.int({ min: 1, max: 100 });
    const descricaoCategoria = `Descricao para a categoria ${categoria}`;

    return {
        nomeProduto: removerAcentos(nomeProduto),
        preco: preco,
        descricaoProduto: removerAcentos(descricao),
        quantidade: quantidade,
        nomeCategoria: removerAcentos(categoria),
        descricaoCategoria: removerAcentos(descricaoCategoria),
    };
}
